README.md
Implementation of Blake encryption algorithm

## Install
```yaml
dependencies:
  blake_hash: ^1.0.2
```

## Usage

A simple usage example:

```dart
import 'dart:typed_data';
import 'package:blake_hash/blake_hash.dart';

void main() {
  Uint8List source = Uint8List.fromList([
    14, 251, 24, 220, 202, 235, 97, 232, //
    85, 229, 109, 188, 53, 146, 43, 36,
    246, 238, 110, 55, 202, 163
  ]);
  Blake blake = Blake256();
  source = blake.update(source).digest();
  print(source);

  /// => [102, 119, 97, 10, 202, 38, 121, 69, 113, 107, 204, 30, 2, 56, 132, 219, 223, 178, 104, 33, 38, 181, 219, 126, 87, 74, 200, 152, 115, 14, 108, 138]

  source = blake.reset().update(source).digest();
  print(source);

  /// => [160, 123, 169, 35, 32, 106, 124, 57, 39, 1, 52, 241, 107, 182, 255, 52, 253, 188, 54, 150, 214, 169, 56, 120, 60, 214, 85, 151, 232, 87, 222, 19]

  blake = Blake512();
  source = blake.update(source).digest();
  print(source);

  /// => [117, 52, 187, 72, 187, 76, 107, 178, 112, 89, 59, 37, 165, 71, 131, 97, 175, 233, 9, 153, 195, 129, 194, 195, 21, 219, 210, 135, 27, 138, 195, 195, 88, 222, 226, 177, 248, 177, 249, 104, 10, 65, 64, 133, 3, 39, 178, 163, 188, 199, 27, 131, 233, 51, 173, 224, 155, 187, 234, 240, 8, 213, 69, 248]

  source = blake.reset().update(source).digest();
  print(source);

  /// => [160, 201, 200, 166, 213, 62, 61, 9, 179, 14, 57, 12, 52, 136, 190, 72, 163, 161, 65, 80, 246, 114, 151, 80, 159, 29, 171, 230, 179, 36, 37, 7, 71, 182, 118, 226, 144, 41, 123, 79, 123, 157, 83, 71, 226, 182, 246, 151, 58, 106, 56, 171, 152, 43, 43, 131, 197, 189, 228, 224, 198, 11, 176, 206]
}
```


## LICENSE

This library is free and open-source software released under the MIT license.


CHANGELOG.md
## 1.0.2

- Support blake512

## 1.0.1

- Support null-safety

## 1.0.0

- Support blake256
- Initial version, created by litao


test/blake_hash_test.dart
import 'dart:typed_data';

import 'package:blake_hash/blake_hash.dart';
import 'package:convert/convert.dart';
import 'package:test/test.dart';

void main() {
  group('blake test', () {
    List<Map<String, String>> vectors = <Map<String, String>>[
      {
        'input': '00',
        'blake256':
            '0ce8d4ef4dd7cd8d62dfded9d4edb0a774ae6a41929a74da23109e8f11139c87',
        'blake512':
            '97961587f6d970faba6d2478045de6d1fabd09b61ae50932054d52bc29d31be4ff9102b9f69e2bbdb83be13d4b9c06091e5fa0b48bd081b634058be0ec49beb3',
      },
      {
        'input':
            '000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
        'blake256':
            'd419bad32d504fb7d44d460c42c5593fe544fa4c135dec31e21bd9abdcc22d41',
        'blake512':
            '2f31e729e269c7f22cac8b71cb80a504afe38350498e0d178afb38cc356dfdaf55d1e7d723c8ff686df0710c43d9abef9f0f61b91ab3b49f72b96f899159a735',
      }
    ];

    vectors.forEach((vector) {
      test('blake256 vector#${vectors.indexOf(vector)}', () {
        expect(
            vector['blake256'],
            hex.encode(Blake256()
                .update(Uint8List.fromList(hex.decode(vector['input']!)))
                .digest()));
      });

      test('blake512 vector#${vectors.indexOf(vector)}', () {
        expect(
            vector['blake512'],
            hex.encode(Blake512()
                .update(Uint8List.fromList(hex.decode(vector['input']!)))
                .digest()));
      });
    });
  });
}


lib/src/blake.dart
part of blake_hash;

Uint8List _zo = Uint8List.fromList([0x01]);
Uint8List _oo = Uint8List.fromList([0x81]);

abstract class Blake {
  static final List<Uint8List> sigma = [
    Uint8List.fromList([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]),
    Uint8List.fromList([14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3]),
    Uint8List.fromList([11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4]),
    Uint8List.fromList([7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8]),
    Uint8List.fromList([9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13]),
    Uint8List.fromList([2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9]),
    Uint8List.fromList([12, 5, 1, 15, 14, 13, 4, 10, 0, 7, 6, 3, 9, 2, 8, 11]),
    Uint8List.fromList([13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10]),
    Uint8List.fromList([6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5]),
    Uint8List.fromList([10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0]),
    Uint8List.fromList([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]),
    Uint8List.fromList([14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3]),
    Uint8List.fromList([11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4]),
    Uint8List.fromList([7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8]),
    Uint8List.fromList([9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13]),
    Uint8List.fromList([2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9])
  ];

  static final List<int> u256 = [
    0x243f6a88, 0x85a308d3, 0x13198a2e, 0x03707344, //
    0xa4093822, 0x299f31d0, 0x082efa98, 0xec4e6c89,
    0x452821e6, 0x38d01377, 0xbe5466cf, 0x34e90c6c,
    0xc0ac29b7, 0xc97c50dd, 0x3f84d5b5, 0xb5470917
  ];

  static final List<int> u512 = [
    ...u256,
    0x9216d5d9, 0x8979fb1b, 0xd1310ba6, 0x98dfb5ac, //
    0x2ffd72db, 0xd01adfb7, 0xb8e1afed, 0x6a267e96,
    0xba7c9045, 0xf12c7f99, 0x24a19947, 0xb3916cf7,
    0x0801f2e2, 0x858efc16, 0x636920d8, 0x71574e69
  ];

  static final Uint8List padding =
      Uint8List.fromList([0x80, ...List.filled(127, 0)]);

  late ByteData _block;
  late int _blockOffset;
  late List<int> _h;
  late List<int> _length;

  void _length_carry(List<int> data) {
    for (int j = 0; j < data.length; ++j) {
      if (data[j] < 0x0100000000) {
        break;
      }
      data[j] -= 0x0100000000;
      data[j + 1] += 1;
    }
  }

  Blake reset();

  void _compress();

  Uint8List digest();

  Blake update(Uint8List data) {
    int offset = 0;

    while (_blockOffset + data.length - offset >= _block.lengthInBytes) {
      for (int i = _blockOffset; i < _block.lengthInBytes;) {
        _block.setUint8(i++, data[offset++]);
      }

      _length[0] += _block.lengthInBytes << 3;
      _length_carry(_length);

      _compress();
      _blockOffset = 0;
    }

    while (offset < data.length) {
      _block.setUint8(_blockOffset++, data[offset++]);
    }
    return this;
  }
}


lib/src/blake256.dart
part of blake_hash;

class Blake256 extends Blake {
  late List<int> _s;
  late bool _nullT;

  Blake256() : super() {
    reset();
  }

  @override
  Blake256 reset() {
    _h = [
      0x6a09e667, 0xbb67ae85, 0x3c6ef372, 0xa54ff53a, //
      0x510e527f, 0x9b05688c, 0x1f83d9ab, 0x5be0cd19
    ];

    _s = List<int>.filled(4, 0);

    _block = ByteData(64);

    _blockOffset = 0;
    _length = List<int>.filled(2, 0);

    _nullT = false;

    return this;
  }

  int _rot(int x, int n) {
    return ((x << (32 - n)) | (x >> n).toUnsigned(32)).toUnsigned(32);
  }

  void _g(List<int> v, List<int> m, int i, int a, int b, int c, int d, int e) {
    List<Uint8List> sigma = Blake.sigma;
    List<int> u256 = Blake.u256;
    v[a] = (v[a] +
            ((m[sigma[i][e]] ^ u256[sigma[i][e + 1]]).toUnsigned(32)) +
            v[b])
        .toUnsigned(32);
    v[d] = _rot(v[d] ^ v[a], 16);
    v[c] = (v[c] + v[d]).toUnsigned(32);
    v[b] = _rot(v[b] ^ v[c], 12);
    v[a] = (v[a] +
            ((m[sigma[i][e + 1]] ^ u256[sigma[i][e]]).toUnsigned(32)) +
            v[b])
        .toUnsigned(32);
    v[d] = _rot(v[d] ^ v[a], 8);
    v[c] = (v[c] + v[d]).toUnsigned(32);
    v[b] = _rot(v[b] ^ v[c], 7);
  }

  @override
  void _compress() {
    List<int> u256 = Blake.u256;
    List<int> v = List.filled(16, 0);
    List<int> m = List.filled(16, 0);

    for (int i = 0; i < 16; ++i) {
      m[i] = _block.getUint32(i * 4);
    }

    for (int i = 0; i < 8; ++i) {
      v[i] = _h[i];
    }
    for (int i = 8; i < 12; ++i) {
      v[i] = (_s[i - 8] ^ u256[i - 8]);
    }
    for (int i = 12; i < 16; ++i) {
      v[i] = u256[i - 8];
    }
    if (!_nullT) {
      v[12] ^= _length[0];
      v[13] ^= _length[0];
      v[14] ^= _length[1];
      v[15] ^= _length[1];
    }

    for (int i = 0; i < 14; ++i) {
      /* column step */
      _g(v, m, i, 0, 4, 8, 12, 0);
      _g(v, m, i, 1, 5, 9, 13, 2);
      _g(v, m, i, 2, 6, 10, 14, 4);
      _g(v, m, i, 3, 7, 11, 15, 6);
      /* diagonal step */
      _g(v, m, i, 0, 5, 10, 15, 8);
      _g(v, m, i, 1, 6, 11, 12, 10);
      _g(v, m, i, 2, 7, 8, 13, 12);
      _g(v, m, i, 3, 4, 9, 14, 14);
    }

    for (int i = 0; i < 16; ++i) {
      _h[i % 8] ^= v[i];
    }
    for (int i = 0; i < 8; ++i) {
      _h[i] ^= _s[i % 4];
    }
  }

  void _padding() {
    int lo = _length[0] + _blockOffset * 8;
    int hi = _length[1];
    if (lo >= 0x0100000000) {
      lo -= 0x0100000000;
      hi += 1;
    }

    ByteData msgLen = ByteData(8);

    msgLen.setUint32(0, hi);

    msgLen.setUint32(4, lo);

    if (_blockOffset == 55) {
      _length[0] -= 8;
      update(_oo);
    } else {
      if (_blockOffset < 55) {
        if (_blockOffset == 0) _nullT = true;
        _length[0] -= (55 - _blockOffset) << 3;
        update(Blake.padding.sublist(0, 55 - _blockOffset));
      } else {
        _length[0] -= (64 - _blockOffset) << 3;
        update(Blake.padding.sublist(0, 64 - _blockOffset));
        _length[0] -= 55 * 8;
        update(Blake.padding.sublist(1, 1 + 55));
        _nullT = true;
      }

      update(_zo);
      _length[0] -= 8;
    }

    _length[0] -= 64;
    update(msgLen.buffer.asUint8List());
  }

  @override
  Uint8List digest() {
    _padding();
    ByteData buffer = ByteData(32);

    for (int i = 0; i < 8; ++i) {
      buffer.setUint32(i << 2, _h[i]);
    }
    return buffer.buffer.asUint8List();
  }
}


lib/src/blake512.dart
part of blake_hash;

class Blake512 extends Blake {
  late List<int> _s;
  late bool _nullT;

  Blake512() : super() {
    reset();
  }

  @override
  Blake512 reset() {
    _h = [
      0x6a09e667, 0xf3bcc908, 0xbb67ae85, 0x84caa73b, //
      0x3c6ef372, 0xfe94f82b, 0xa54ff53a, 0x5f1d36f1,
      0x510e527f, 0xade682d1, 0x9b05688c, 0x2b3e6c1f,
      0x1f83d9ab, 0xfb41bd6b, 0x5be0cd19, 0x137e2179
    ];

    _s = List<int>.filled(8, 0);

    _block = ByteData(128);

    _blockOffset = 0;
    _length = List<int>.filled(4, 0);

    _nullT = false;

    return this;
  }

  void _rot(List<int> v, int i, int j, int n) {
    int hi = v[i * 2] ^ v[j * 2];
    int lo = v[i * 2 + 1] ^ v[j * 2 + 1];

    if (n >= 32) {
      lo = lo ^ hi;
      hi = lo ^ hi;
      lo = lo ^ hi;
      n -= 32;
    }

    if (n == 0) {
      v[i * 2] = hi.toUnsigned(32);
      v[i * 2 + 1] = lo.toUnsigned(32);
    } else {
      v[i * 2] = ((hi >> n).toUnsigned(32) | (lo << (32 - n))).toUnsigned(32);
      v[i * 2 + 1] =
          ((lo >> n).toUnsigned(32) | (hi << (32 - n))).toUnsigned(32);
    }
  }

  void _g(List<int> v, List<int> m, int i, int a, int b, int c, int d, int e) {
    List<Uint8List> sigma = Blake.sigma;
    List<int> u512 = Blake.u512;
    int lo = v[a * 2 + 1] +
        ((m[sigma[i][e] * 2 + 1] ^ u512[sigma[i][e + 1] * 2 + 1])
            .toUnsigned(32)) +
        v[b * 2 + 1];
    v[a * 2] = (v[a * 2] +
            ((m[sigma[i][e] * 2] ^ u512[sigma[i][e + 1] * 2]).toUnsigned(32)) +
            v[b * 2] +
            (lo ~/ 0x0100000000))
        .toUnsigned(32);
    v[a * 2 + 1] = lo.toUnsigned(32);
    _rot(v, d, a, 32);
    lo = v[c * 2 + 1] + v[d * 2 + 1];
    v[c * 2] = (v[c * 2] + v[d * 2] + (lo ~/ 0x0100000000)).toUnsigned(32);
    v[c * 2 + 1] = lo.toUnsigned(32);
    _rot(v, b, c, 25);
    lo = v[a * 2 + 1] +
        ((m[sigma[i][e + 1] * 2 + 1] ^ u512[sigma[i][e] * 2 + 1])
            .toUnsigned(32)) +
        v[b * 2 + 1];
    v[a * 2] = (v[a * 2] +
            ((m[sigma[i][e + 1] * 2] ^ u512[sigma[i][e] * 2]).toUnsigned(32)) +
            v[b * 2] +
            (lo ~/ 0x0100000000))
        .toUnsigned(32);
    v[a * 2 + 1] = lo.toUnsigned(32);
    _rot(v, d, a, 16);
    lo = v[c * 2 + 1] + v[d * 2 + 1];
    v[c * 2] = (v[c * 2] + v[d * 2] + (lo ~/ 0x0100000000)).toUnsigned(32);
    v[c * 2 + 1] = lo.toUnsigned(32);
    _rot(v, b, c, 11);
  }

  @override
  void _compress() {
    List<int> u512 = Blake.u512;
    List<int> v = List<int>.filled(32, 0);
    List<int> m = List<int>.filled(32, 0);
    int i = 0;
    for (; i < 32; ++i) {
      m[i] = _block.getUint32(i << 2);
    }
    for (i = 0; i < 16; ++i) {
      v[i] = _h[i].toUnsigned(32);
    }
    for (i = 16; i < 24; ++i) {
      v[i] = (_s[i - 16] ^ u512[i - 16]).toUnsigned(32);
    }
    for (i = 24; i < 32; ++i) {
      v[i] = u512[i - 16];
    }
    if (!_nullT) {
      v[24] = (v[24] ^ _length[1]).toUnsigned(32);
      v[25] = (v[25] ^ _length[0]).toUnsigned(32);
      v[26] = (v[26] ^ _length[1]).toUnsigned(32);
      v[27] = (v[27] ^ _length[0]).toUnsigned(32);
      v[28] = (v[28] ^ _length[3]).toUnsigned(32);
      v[29] = (v[29] ^ _length[2]).toUnsigned(32);
      v[30] = (v[30] ^ _length[3]).toUnsigned(32);
      v[31] = (v[31] ^ _length[2]).toUnsigned(32);
    }
    for (i = 0; i < 16; ++i) {
      _g(v, m, i, 0, 4, 8, 12, 0);
      _g(v, m, i, 1, 5, 9, 13, 2);
      _g(v, m, i, 2, 6, 10, 14, 4);
      _g(v, m, i, 3, 7, 11, 15, 6);
      _g(v, m, i, 0, 5, 10, 15, 8);
      _g(v, m, i, 1, 6, 11, 12, 10);
      _g(v, m, i, 2, 7, 8, 13, 12);
      _g(v, m, i, 3, 4, 9, 14, 14);
    }

    for (i = 0; i < 16; ++i) {
      _h[(i % 8) * 2] = (_h[(i % 8) * 2] ^ v[i * 2]).toUnsigned(32);
      _h[(i % 8) * 2 + 1] = (_h[(i % 8) * 2 + 1] ^ v[i * 2 + 1]).toUnsigned(32);
    }

    for (i = 0; i < 8; ++i) {
      _h[i * 2] = (_h[i * 2] ^ _s[(i % 4) * 2]).toUnsigned(32);
      _h[i * 2 + 1] = (_h[i * 2 + 1] ^ _s[(i % 4) * 2 + 1]).toUnsigned(32);
    }
  }

  void _padding() {
    List<int> len = _length.sublist(0);
    len[0] += _blockOffset << 3;
    _length_carry(len);
    ByteData msgLen = ByteData(16);
    for (int i = 0; i < 4; ++i) {
      msgLen.setUint32(i << 2, len[3 - i]);
    }
    if (_blockOffset == 111) {
      _length[0] -= 8;
      update(_oo);
    } else {
      if (_blockOffset < 111) {
        if (_blockOffset == 0) {
          _nullT = true;
        }
        _length[0] -= (111 - _blockOffset) << 3;
        update(Blake.padding.sublist(0, 111 - _blockOffset));
      } else {
        _length[0] -= (128 - _blockOffset) << 3;
        update(Blake.padding.sublist(0, 128 - _blockOffset));
        _length[0] -= 888;
        update(Blake.padding.sublist(1, 112));
        _nullT = true;
      }
      update(_zo);
      _length[0] -= 8;
    }
    _length[0] -= 128;
    update(msgLen.buffer.asUint8List());
  }

  @override
  Uint8List digest() {
    _padding();
    ByteData buffer = ByteData(64);
    for (int i = 0; i < 16; ++i) {
      buffer.setUint32(i << 2, _h[i]);
    }
    return buffer.buffer.asUint8List();
  }
}


example/blake_hash_example.dart
import 'dart:typed_data';
import 'package:blake_hash/blake_hash.dart';

void main() {
  Uint8List source = Uint8List.fromList([
    14, 251, 24, 220, 202, 235, 97, 232, //
    85, 229, 109, 188, 53, 146, 43, 36,
    246, 238, 110, 55, 202, 163
  ]);
  Blake blake = Blake256();
  source = blake.update(source).digest();
  print(source);

  /// => [102, 119, 97, 10, 202, 38, 121, 69, 113, 107, 204, 30, 2, 56, 132, 219, 223, 178, 104, 33, 38, 181, 219, 126, 87, 74, 200, 152, 115, 14, 108, 138]

  source = blake.reset().update(source).digest();
  print(source);

  /// => [160, 123, 169, 35, 32, 106, 124, 57, 39, 1, 52, 241, 107, 182, 255, 52, 253, 188, 54, 150, 214, 169, 56, 120, 60, 214, 85, 151, 232, 87, 222, 19]

  blake = Blake512();
  source = blake.update(source).digest();
  print(source);

  /// => [117, 52, 187, 72, 187, 76, 107, 178, 112, 89, 59, 37, 165, 71, 131, 97, 175, 233, 9, 153, 195, 129, 194, 195, 21, 219, 210, 135, 27, 138, 195, 195, 88, 222, 226, 177, 248, 177, 249, 104, 10, 65, 64, 133, 3, 39, 178, 163, 188, 199, 27, 131, 233, 51, 173, 224, 155, 187, 234, 240, 8, 213, 69, 248]

  source = blake.reset().update(source).digest();
  print(source);

  /// => [160, 201, 200, 166, 213, 62, 61, 9, 179, 14, 57, 12, 52, 136, 190, 72, 163, 161, 65, 80, 246, 114, 151, 80, 159, 29, 171, 230, 179, 36, 37, 7, 71, 182, 118, 226, 144, 41, 123, 79, 123, 157, 83, 71, 226, 182, 246, 151, 58, 106, 56, 171, 152, 43, 43, 131, 197, 189, 228, 224, 198, 11, 176, 206]
}
